### PR TITLE
kernel: need to release spinlock before busy_wait

### LINKED ISSUE #20045: z_sched_abort: sched_spinlock should be released before k_busy_wait
**Describe the bug**
In z_sched_abort of thread_abort.c
```
LOCKED(&sched_spinlock) {
			if (z_is_thread_prevented_from_running(thread)) {
				__ASSERT(!z_is_thread_queued(thread), "");
				thread->base.thread_state |= _THREAD_DEAD;
			} else if (z_is_thread_queued(thread)) {
				_priq_run_remove(&_kernel.ready_q.runq, thread);
				z_mark_thread_as_not_queued(thread);
				thread->base.thread_state |= _THREAD_DEAD;
			} else {
				k_busy_wait(100);
			}
		}
```
the sched_spinlock should be released before k_busy_wait(100),  if not released,  other cores cannot get the sched_spinlock during k_busy_wait(100).  This will worse the compete of locks, especially for cores >= 4

### PR DESCRIPTION
**Pull request by vonhust**
need to release spinlock first before busy_wait,
or other cores cannot get the spinlock when the holder is
busy waitting.

Fixes #20045

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -1065,17 +1068,19 @@ void z_sched_abort(struct k_thread *thread)
...
+			thread->base.thread_state |= _THREAD_DEAD;
+			k_spin_unlock(&sched_spinlock, key);
+		} else if (z_is_thread_queued(thread)) {
+			_priq_run_remove(&_kernel.ready_q.runq, thread);
+			z_mark_thread_as_not_queued(thread);
+			thread->base.thread_state |= _THREAD_DEAD;
+			k_spin_unlock(&sched_spinlock, key);
+		} else {
+			k_spin_unlock(&sched_spinlock, key);
+			k_busy_wait(100);
```
Comment (andyross) [inline kernel/sched.c]: Style wise, I prefer to keep the lock and unlock pairs at the same level of indentation.  Can you change this to something like:

```c
    k_spin_unlock(&sched_spinlock, key);
    k_busy_wait(100);
    k_spin_lock(&sched_spinlock);
```

File: kernel/sched.c [hunk 2]
```diff
@@ -1047,6 +1047,9 @@ void z_sched_ipi(void)
 
 void z_sched_abort(struct k_thread *thread)
 {
+
```
Comment (andyross) [inline kernel/sched.c]: Pedantry: needless blank line

**Review by andyross — COMMENTED** [review]
Good catch.  Not loving the structure though, with the lock taken at one level and then released independently in each case in a chained if.  Can you go back to a single outer "lock/unlock" pair, and just "unlock/lock" around the busy wait?

**Comment by andyross** [comment]
Oh, and do you have a test case that caught this latency problem?  IIRC, when I added it I went with the simple busy wait because I couldn't actually get the spin to happen.  The IPI on qemu, at least, seems to happen almost synchronously and I couldn't catch it before it completed.

**Comment by vonhust** [comment]
> Oh, and do you have a test case that caught this latency problem? IIRC, when I added it I went with the simple busy wait because I couldn't actually get the spin to happen. The IPI on qemu, at least, seems to happen almost synchronously and I couldn't catch it before it completed.

I found this in debugging of ARC HSDK which has 4 cores, using tests/kernel/smp

File: kernel/sched.c (hunk 1 above)
Comment (vonhust) [reply kernel/sched.c]: I did't find a way to keep the lock and unlock parirs at the same level.  In the provided codes, k_spin_lock cannot be after k_busy_wait  as while((thread->base.thread & _THREAD_DEAD)) may exit before call k_spin_unlock.   How about the following:

```
		LOCKED(&sched_spinlock) {
			if (z_is_thread_prevented_from_running(thread)) {
				__ASSERT(!z_is_thread_queued(thread), "");
				thread->base.thread_state |= _THREAD_DEAD;
			} else if (z_is_thread_queued(thread)) {
				_priq_run_remove(&_kernel.ready_q.runq, thread);
				z_mark_thread_as_not_queued(thread);
				thread->base.thread_state |= _THREAD_DEAD;
			}
		}
                k_busy_wait(100);
```
But the problem is that k_busy_wait(100) will be always executed

File: kernel/sched.c (hunk 2 above)
Comment (vonhust) [reply kernel/sched.c]: Fixed

**Review by ruuddw — APPROVED** [review]
@andyross : this is blocking for ARC SMP work, can you please approve or suggest another way for handling the lock/inlock indentation? Thanks.

**Review by andyross — APPROVED** [review]
OK, this seems better, thanks.

**Pull request merged**